### DISCUSSION TITLE
Why formState is not updated when using useImperativeHandle and testing with react-testing-library

### DISCUSSION THREAD

**Discussion by einatnielsen** [react-hook-form/react-hook-form#2582] [Q&A]
I have written the following component with 2 inputs, where one of them is required.
I have a special scenario where I need to trigger the submit from a parent component, so I used useImperativeHandle to expose the submit.
When I run a test with the required input as empty, I expect the formState.isValid to be false (after running handleSubmit) , but instead its true.
Has anyone encountered this issue? is this a bug? I am guessing it happens because I am using `useImperativeHandle`. I just dont know why formState is not updated when running the test. I tried to wrap the fireEvent that triggers the onSave with act, it didnt help

```

interface MyComponentProps {
  name: string
  description: string
  onSave: Function
}

const MyComponent = React.forwardRef((props: MyComponentProps, ref) => {
  const { register, handleSubmit, formState, errors, setValue } = useForm({
    mode: 'onChange',
    defaultValues: {
        'name': props.name,
        'description': props.description
      }
  })

  React.useImperativeHandle(ref, () => ({
    onSave: () => {
      handleSubmit(handleSave)()
      return formState.isValid
    }
  }))

  const handleSave = formFields => {
    props.onSave({name: formFields['name'], description: formFields['description']})
  }

  React.useEffect(() => {
      setValue('name', props.name, true)
      setValue('description', description)
  }, [props.name, props.description])

   return {
      <div>
        <input name='name' ref={register({
            required: true
          })}/>
        <input name='description' ref={register} />
      </div>
  }
})

```

Thanks!!

**Comment by bluebill1049** [comment]
pretty sure it's due to proxy: https://react-hook-form.com/api#formState

just need to read the value before `render`

**Reply by einatnielsen** [reply]
thanks @bluebill1049 , but I am not sure I understand, can you please explain it with a code example?

**Reply by bluebill1049** [reply]
Have a look here: https://react-hook-form.com/api#formState there is an important note.

```
import React from "react";
import { useForm } from "react-hook-form";

export default function App() {
  const { register, handleSubmit, errors, formState } = useForm();
  const onSubmit = data => console.log(data);
  
  // Read the formState before render to subscribe the form state through Proxy
  const { isDirty, isSubmitting, touched, submitCount } = formState; // see it's been invoked here, which means those formState is subscribed.
  
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input name="test" ref={register} />
      <input type="submit" />
    </form>
  );
}
```

**Comment by einatnielsen** [comment]
@bluebill1049 thanks for the reply, I did what you suggested but when I debug my test, isValid is still true, while it should be false

**Reply by bluebill1049** [reply]
can you reproduce this in a codesandbox?

**Comment by inPhoenix** [comment]
Hi everyone.

First I would thank you for the amazing library.

I noticed the same issue when performing onChange events using enzyme for a wrapper component using [Controller](https://react-hook-form.com/v6/api#Controller).

My test code is similar to:

```javascript

  const wrapper = shallow(<SignUpForm />)
  
  const textBoxField = UI.find('[name="myField"]').props()

  await act(async () => {
    textBoxField .onChange('text') // this will not change the formState
    textBoxField .onBlur()
  })
  wrapper.update()
 
  expect(wrapper.find('[name="email"]').prop('error')).toBe(true)

```

My componentt code is similar to:

```javascript

const handleOnBlur = (onBlur) => () => {
  const { isDirty } = formState
    if (isDirty) {
      onBlur()
    }
    console('isDitry?', isDirty) // true for real input and false for test onChange event
}

<Controller
      render={(props) => (
        <MyTextBox
          error={errors.myField}
          onBlur={handleOnBlur(props.onBlur)}
          // ... other props as onChange that also updates the state etc...
        />
      )}
      name="myField"
      control={control}
      rules={{
        required: 'should not be empty',
        validate: (value) => {
          if (!value.trim()) {
            return 'should not be empty'
          }
        }
      }}
    />         
```

So basically I have a onBlurHandler that checks if the formState is dirty _(isDirty)_ to trigger the props.onBlur() only if the field is dirty. However, it works well on the App it does not change the form for in the test for onChange event: textBoxField .onChange('text').

So anyone knows any tricky to do that or have workaround similar behaviour? I tried to provoke more onChange events and update() events but without success.
